lsp-mode.el

```diff
@@ -3794,7 +3794,7 @@ PLIST is the additional data to attach to each candidate."
                   (prefix-line (buffer-substring-no-properties (point-at-bol) (point))))
             (setf done? (or (seqp resp)
                             (not (gethash "isIncomplete" resp)))
-                  lsp--capf-cache (when (and done? (car bounds))
+                  lsp--capf-cache (when (and done? (car bounds) (not (seq-empty-p items)))
```

Comment:
Do we need `(car bounds)` here?

Reply:
Yes, just in case the bounds is `nil` or we cannot get the start point of current symbol.
In that case, we don't have the prefix to make sure that the cache is for current input and should not risk caching it.

Reply:
Isn't prefix-line + start point + cache cleanup sufficient? Otherwise, we won't cache after trigger character and when triggering the completion with no symbol. AFAICS after  `(car bounds)` is put in the cache it is not used later.

Reply:
The `bounds` is used to calculate the prefix for caching. That may different from the prefix we used for filtering.
The reason is to get the prefix for filtering, we actually have to wait for server response (textEdit), whereas the bounds can be calculated directly from `thing-at-point`.
And we put prefix, not `(car bounds)` into cache.

Reply:
Hm, why do we use `(buffer-substring-no-properties (car bounds) (point))` to calculate the prefix but not `(buffer-substring-no-properties start (point))` and later reuse that `start` point if the cache prefix-line is a prefix of current prefix-line? Am I missing something? Also, seems like we should default bounds to `(list (point) (point))` when symbol at point returns nil. This will also remove the need from checking for `(car bounds)` which will be guaranteed to be not nil.

Reply:
The prefix here is used just for cache checking purpose. When we do filtering, we're using cached `start-point`.
We can certainly use `prefix-line` instead of prefix for cache checking. But I doubt it would be perf hit when the line is long. So it's more like the longer we take, the slower completion become even with cache, which is not good.

Bound `bounds` to default value also good for me, will update that.

Reply:
> That means the cache is not working for jdtls, I can only imagine that they never send `isIncomplete` with `nil`

Not cache - I do not get any results when I do `(setq completion-styles '(flex))` but it works with default `completion-styles`